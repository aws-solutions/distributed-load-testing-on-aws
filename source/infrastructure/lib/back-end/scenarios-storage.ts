// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { RemovalPolicy } from "aws-cdk-lib";
import { AttributeType, BillingMode, Table, TableEncryption } from "aws-cdk-lib/aws-dynamodb";
import { AnyPrincipal, Effect, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { BlockPublicAccess, Bucket, BucketEncryption, HttpMethods } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

export interface ScenarioTestRunnerStorageConstructProps {
  // S3 Logs Bucket
  readonly s3LogsBucket: Bucket;
  // CloudFront domain name
  readonly cloudFrontDomainName: string;
  // Solution Id
  readonly solutionId: string;
}

/**
 * Distributed Load Testing storage construct
 * Creates an S3 bucket to store test scenarios and
 * a Dynamodb table to store tests and test configuration
 */
export class ScenarioTestRunnerStorageConstruct extends Construct {
  public scenariosBucket: Bucket;
  public scenariosS3Policy: Policy;
  public scenariosTable: Table;
  public scenarioDynamoDbPolicy: Policy;
  public historyTable: Table;
  public historyDynamoDbPolicy: Policy;

  constructor(scope: Construct, id: string, props: ScenarioTestRunnerStorageConstructProps) {
    super(scope, id);

    this.scenariosBucket = new Bucket(this, "DLTScenariosBucket", {
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: props.s3LogsBucket,
      serverAccessLogsPrefix: "scenarios-bucket-access/",
      encryption: BucketEncryption.KMS_MANAGED,
      enforceSSL: true,
      versioned: true,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      cors: [
        {
          allowedMethods: [HttpMethods.GET, HttpMethods.POST, HttpMethods.PUT],
          allowedOrigins: [`https://${props.cloudFrontDomainName}`],
          allowedHeaders: ["*"],
          exposedHeaders: ["ETag"],
        },
      ],
    });

    this.scenariosBucket.addToResourcePolicy(
      new PolicyStatement({
        actions: ["s3:*"],
        resources: [this.scenariosBucket.bucketArn, `${this.scenariosBucket.bucketArn}/*`],
        effect: Effect.DENY,
        principals: [new AnyPrincipal()],
        conditions: {
          Bool: {
            "aws:SecureTransport": false,
          },
        },
      })
    );

    this.scenariosS3Policy = new Policy(this, "ScenariosS3Policy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:DeleteObject", "s3:ListBucket"],
          resources: [this.scenariosBucket.bucketArn, `${this.scenariosBucket.bucketArn}/*`],
        }),
      ],
    });

    this.scenariosTable = new Table(this, "DLTScenariosTable", {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: "testId", type: AttributeType.STRING },
      pointInTimeRecovery: true,
    });

    this.historyTable = new Table(this, "DLTHistoryTable", {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: "testId", type: AttributeType.STRING },
      sortKey: { name: "testRunId", type: AttributeType.STRING },
      pointInTimeRecovery: true,
    });

    const historyDDBActions = ["dynamodb:BatchWriteItem", "dynamodb:PutItem", "dynamodb:Query"];
    this.historyDynamoDbPolicy = new Policy(this, "HistoryDynamoDbPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: historyDDBActions,
          resources: [this.historyTable.tableArn],
        }),
      ],
    });

    const scenariosDDBActions = [
      "dynamodb:DeleteItem",
      "dynamodb:GetItem",
      "dynamodb:PutItem",
      "dynamodb:Scan",
      "dynamodb:UpdateItem",
    ];
    this.scenarioDynamoDbPolicy = new Policy(this, "ScenarioDynamoDbPolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: scenariosDDBActions,
          resources: [this.scenariosTable.tableArn],
        }),
      ],
    });
  }
}
