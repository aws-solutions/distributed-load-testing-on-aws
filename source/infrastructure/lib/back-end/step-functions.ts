// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  Chain,
  Choice,
  Condition,
  Fail,
  Pass,
  LogLevel,
  Map as SFMap,
  StateMachine,
  Succeed,
  Wait,
  WaitTime,
  JsonPath,
  DefinitionBody,
  TaskInput,
} from "aws-cdk-lib/aws-stepfunctions";
import { LambdaInvoke } from "aws-cdk-lib/aws-stepfunctions-tasks";
import { Aws, CfnResource, Duration } from "aws-cdk-lib";
import { Policy } from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { IFunction } from "aws-cdk-lib/aws-lambda";
import { Construct } from "constructs";

export interface TaskRunnerStepFunctionConstructProps {
  // State machine Lambda functions
  readonly taskStatusChecker: IFunction;
  readonly taskRunner: IFunction;
  readonly resultsParser: IFunction;
  readonly taskCanceler: IFunction;
  // Solution ID
  readonly solutionId: string;
  readonly suffix: string;
}

/**
 * Creates the Step function state machine to control the Fargate tasks
 */
export class TaskRunnerStepFunctionConstruct extends Construct {
  public taskRunnerStepFunctions: StateMachine;

  constructor(scope: Construct, id: string, props: TaskRunnerStepFunctionConstructProps) {
    super(scope, id);

    const stepFunctionsLogGroup = new LogGroup(this, "StepFunctionsLogGroup", {
      retention: RetentionDays.ONE_YEAR,
      logGroupName: `/aws/vendedlogs/states/StepFunctionsLogGroup${Aws.STACK_NAME}${props.suffix}`,
    });
    const stepFunctionsLogGroupResource = stepFunctionsLogGroup.node.defaultChild as CfnResource;
    stepFunctionsLogGroupResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });

    const done = new Succeed(this, "Done");
    const mapEnd = new Pass(this, "Map End");
    const parseResult = new LambdaInvoke(this, "Parse result", {
      lambdaFunction: props.resultsParser,
      payload: TaskInput.fromObject({
        "testTaskConfig.$": "$.testTaskConfig",
        "testId.$": "$.testId",
        "testType.$": "$.testType",
        "fileType.$": "$.fileType",
        "showLive.$": "$.showLive",
        "testDuration.$": "$.testDuration",
        "prefix.$": "$.prefix",
        "executionStart.$": "$$.Execution.StartTime",
      }),
    });
    parseResult.next(done);

    const checkWorkerStatus = new LambdaInvoke(this, "Check worker status", {
      lambdaFunction: props.taskStatusChecker,
      inputPath: "$",
      outputPath: "$.Payload",
    });

    const checkTaskStatus = new LambdaInvoke(this, "Check task status", {
      lambdaFunction: props.taskStatusChecker,
      inputPath: "$",
      outputPath: "$.Payload",
    });

    const waitTask = new Wait(this, "Wait 1 minute - task status", {
      comment: "Wait 1 minute to check task status again",
      time: WaitTime.duration(Duration.seconds(60)),
    });
    waitTask.next(checkTaskStatus);

    const waitTestDuration = new Wait(this, "Wait specified test duration", {
      time: WaitTime.secondsPath("$.testDuration"),
    });
    waitTestDuration.next(checkTaskStatus);

    const allTasksDone = new Choice(this, "Are all tasks done?");
    allTasksDone.when(Condition.booleanEquals("$.isRunning", false), mapEnd);
    allTasksDone.otherwise(waitTask);

    checkTaskStatus.next(allTasksDone);

    const cancelTest = new LambdaInvoke(this, "Cancel Test", {
      lambdaFunction: props.taskCanceler,
      inputPath: "$",
      resultPath: JsonPath.DISCARD,
    });
    cancelTest.next(mapEnd);

    const waitWorker = new Wait(this, "Wait 1 minute - worker status", {
      comment: "Wait 1 minute to check task status again",
      time: WaitTime.duration(Duration.seconds(60)),
    });
    waitWorker.next(checkWorkerStatus);

    const regionConfigsForTest = new SFMap(this, "Regions for testing", {
      inputPath: "$",
      resultPath: JsonPath.DISCARD,
      itemsPath: "$.testTaskConfig",
      itemSelector: {
        "testTaskConfig.$": "$$.Map.Item.Value",
        "testId.$": "$.testId",
        "testType.$": "$.testType",
        "fileType.$": "$.fileType",
        "showLive.$": "$.showLive",
        "testDuration.$": "$.testDuration",
        "prefix.$": "$.prefix",
      },
    });

    const runWorkers = new LambdaInvoke(this, "Run workers", {
      lambdaFunction: props.taskRunner,
      inputPath: "$",
      outputPath: "$.Payload",
    });

    const requiresLeader = new Choice(this, "Requires leader?");
    requiresLeader.when(Condition.booleanEquals("$.isRunning", false), cancelTest);
    requiresLeader.when(Condition.isNotPresent("$.taskIds"), waitTestDuration);
    requiresLeader.otherwise(waitWorker);

    runWorkers.next(requiresLeader);

    const runLeaderTask = new LambdaInvoke(this, "Run leader task", {
      lambdaFunction: props.taskRunner,
      inputPath: "$",
      outputPath: "$.Payload",
    });
    runLeaderTask.addCatch(cancelTest, { resultPath: "$.error" });

    runLeaderTask.next(waitTestDuration);

    const allWorkersRunning = new Choice(this, "Are all workers running?");
    allWorkersRunning.when(Condition.booleanEquals("$.isRunning", false), cancelTest);
    allWorkersRunning.when(Condition.numberEqualsJsonPath("$.numTasksRunning", "$.numTasksTotal"), runLeaderTask);
    allWorkersRunning.otherwise(waitWorker);

    checkWorkerStatus.next(allWorkersRunning);

    const testIsStillRunning = new Fail(this, "Test is still running", {
      cause: "The same test is already running.",
      error: "TestAlreadyRunning",
    });

    const noRunningTests = new Choice(this, "No running tests");
    noRunningTests.when(Condition.booleanEquals("$.isRunning", false), runWorkers);
    noRunningTests.otherwise(testIsStillRunning);

    const checkRunningTests = new LambdaInvoke(this, "Check running tests", {
      lambdaFunction: props.taskStatusChecker,
      inputPath: "$",
      outputPath: "$.Payload",
    });
    checkRunningTests.next(noRunningTests);

    const definition = Chain.start(regionConfigsForTest.itemProcessor(checkRunningTests)).next(parseResult);

    this.taskRunnerStepFunctions = new StateMachine(this, "TaskRunnerStepFunctions", {
      definitionBody: DefinitionBody.fromChainable(definition),
      logs: {
        destination: stepFunctionsLogGroup,
        level: LogLevel.ALL,
        includeExecutionData: false,
      },
    });
    const stepFunctionsRoleResource = this.taskRunnerStepFunctions.role.node.defaultChild as CfnResource;
    stepFunctionsRoleResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "CloudWatch logs actions do not support resource level permissions",
        },
        {
          id: "W12",
          reason: "CloudWatch logs actions do not support resource level permissions",
        },
      ],
    });
    const stepFunctionPolicy = this.taskRunnerStepFunctions.role.node.findChild("DefaultPolicy") as Policy;
    const policyResource = stepFunctionPolicy.node.defaultChild as CfnResource;
    policyResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W12",
          reason: "CloudWatch logs actions do not support resource level permissions",
        },
        {
          id: "W76",
          reason: "The IAM policy is written for least-privilege access.",
        },
      ],
    });
  }
}
