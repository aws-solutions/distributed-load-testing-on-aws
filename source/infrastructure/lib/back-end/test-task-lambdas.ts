// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, Aws, CfnResource, Duration, Stack } from "aws-cdk-lib";
import { Code, Function as LambdaFunction, Runtime } from "aws-cdk-lib/aws-lambda";
import { Effect, PolicyStatement, PolicyDocument, Role, ServicePrincipal, Policy } from "aws-cdk-lib/aws-iam";
import { IBucket } from "aws-cdk-lib/aws-s3";
import { Table } from "aws-cdk-lib/aws-dynamodb";
import { LogGroup } from "aws-cdk-lib/aws-logs";
import { Construct } from "constructs";

export interface TestRunnerLambdasConstructProps {
  readonly cloudWatchLogsPolicy: Policy;
  // ECS Task Execution Role ARN
  readonly ecsTaskExecutionRoleArn: string;
  // ECS CloudWatch LogGroup
  readonly ecsCloudWatchLogGroup: LogGroup;
  // ECS Cluster
  readonly ecsCluster: string;
  // ECS Task definition
  readonly ecsTaskDefinition: string;
  // ECS Security Group
  readonly ecsTaskSecurityGroup: string;
  // Table storing historical test data
  readonly historyTable: Table;
  // History DynamoDB table policy
  readonly historyDynamoDbPolicy: Policy;
  // Scenarios S3 Bucket policy
  readonly scenariosS3Policy: Policy;
  // Subnet A Id
  readonly subnetA: string;
  // Subnet B Id
  readonly subnetB: string;
  // Test scenarios bucket
  readonly scenariosBucket: string;
  // Test scenarios table
  readonly scenariosTable: Table;
  // Scenario DynamoDB table policy
  readonly scenariosDynamoDbPolicy: Policy;
  /**
   * Solution config properties.
   * the metric URL endpoint, send anonymized usage, solution ID, version, source code bucket, and source code prefix
   */
  readonly metricsUrl: string;
  readonly sendAnonymizedUsage: string;
  readonly solutionId: string;
  readonly solutionVersion: string;
  readonly sourceCodeBucket: IBucket;
  readonly sourceCodePrefix: string;
  readonly uuid: string;
}

/**
 * Distributed Load Testing on AWS Test Runner Lambdas construct.
 * This creates the Results parser, Task Runner, Task Canceler,
 * and Task Status Checker
 */
export class TestRunnerLambdasConstruct extends Construct {
  public resultsParser: LambdaFunction;
  public taskRunner: LambdaFunction;
  public taskCanceler: LambdaFunction;
  public taskCancelerInvokePolicy: Policy;
  public taskStatusChecker: LambdaFunction;
  public realTimeDataPublisher: LambdaFunction;

  constructor(scope: Construct, id: string, props: TestRunnerLambdasConstructProps) {
    super(scope, id);

    const ecsLogGroupArn = props.ecsCloudWatchLogGroup.logGroupArn;
    const lambdaResultsRole = new Role(this, "LambdaResultsRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
    });
    const cfnPolicy = new Policy(this, "LambdaResultsPolicy", {
      statements: [
        new PolicyStatement({
          resources: ["*"],
          actions: ["cloudwatch:GetMetricWidgetImage"],
        }),
        new PolicyStatement({
          resources: [ecsLogGroupArn],
          actions: ["logs:DeleteMetricFilter"],
        }),
      ],
    });

    lambdaResultsRole.attachInlinePolicy(cfnPolicy);
    lambdaResultsRole.attachInlinePolicy(props.cloudWatchLogsPolicy);
    lambdaResultsRole.attachInlinePolicy(props.scenariosDynamoDbPolicy);
    lambdaResultsRole.attachInlinePolicy(props.historyDynamoDbPolicy);
    lambdaResultsRole.attachInlinePolicy(props.scenariosS3Policy);

    const resultsRoleResource = lambdaResultsRole.node.defaultChild as CfnResource;
    resultsRoleResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W12",
          reason: "The action does not support resource level permissions.",
        },
      ],
    });
    const resultsPolicyResource = cfnPolicy.node.defaultChild as CfnResource;
    resultsPolicyResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W12",
          reason: "The action does not support resource level permissions.",
        },
      ],
    });

    this.resultsParser = new LambdaFunction(this, "ResultsParser", {
      description: "Result parser for indexing xml test results to DynamoDB",
      handler: "index.handler",
      role: lambdaResultsRole,
      code: Code.fromBucket(props.sourceCodeBucket, `${props.sourceCodePrefix}/results-parser.zip`),
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(120),
      environment: {
        HISTORY_TABLE: props.historyTable.tableName,
        METRIC_URL: props.metricsUrl,
        SCENARIOS_BUCKET: props.scenariosBucket,
        SCENARIOS_TABLE: props.scenariosTable.tableName,
        SEND_METRIC: props.sendAnonymizedUsage,
        SOLUTION_ID: props.solutionId,
        UUID: props.uuid,
        VERSION: props.solutionVersion,
      },
    });
    const resultsParserResource = this.resultsParser.node.defaultChild as CfnResource;
    resultsParserResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "This Lambda function does not require a VPC",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });

    const taskArn = Stack.of(this).formatArn({
      service: "ecs",
      resource: "task",
      resourceName: "*",
      arnFormat: ArnFormat.SLASH_RESOURCE_NAME,
    });
    const taskDefArn = Stack.of(this).formatArn({ service: "ecs", resource: "task-definition", resourceName: "*:*" });

    const lambdaTaskRole = new Role(this, "DLTTestLambdaTaskRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        TaskLambdaPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:ListTasks"],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:RunTask", "ecs:DescribeTasks"],
              resources: [taskArn, taskDefArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iam:PassRole"],
              resources: [props.ecsTaskExecutionRoleArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["logs:PutMetricFilter"],
              resources: [ecsLogGroupArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["cloudwatch:PutDashboard"],
              resources: [`arn:${Aws.PARTITION}:cloudwatch::${Aws.ACCOUNT_ID}:dashboard/EcsLoadTesting*`],
            }),
          ],
        }),
      },
    });
    lambdaTaskRole.attachInlinePolicy(props.cloudWatchLogsPolicy);
    lambdaTaskRole.attachInlinePolicy(props.scenariosDynamoDbPolicy);

    const lambdaTaskRoleResource = lambdaTaskRole.node.defaultChild as CfnResource;
    lambdaTaskRoleResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "ecs:ListTasks does not support resource level permissions",
        },
      ],
    });

    this.taskRunner = new LambdaFunction(this, "TaskRunner", {
      description: "Task runner for ECS task definitions",
      handler: "index.handler",
      role: lambdaTaskRole,
      code: Code.fromBucket(props.sourceCodeBucket, `${props.sourceCodePrefix}/task-runner.zip`),
      environment: {
        SCENARIOS_BUCKET: props.scenariosBucket,
        SCENARIOS_TABLE: props.scenariosTable.tableName,
        SOLUTION_ID: props.solutionId,
        VERSION: props.solutionVersion,
      },
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(900),
    });
    const taskRunnerResource = this.taskRunner.node.defaultChild as CfnResource;
    taskRunnerResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "This Lambda function does not require a VPC",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });

    const taskCancelerRole = new Role(this, "LambdaTaskCancelerRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        TaskCancelerPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:ListTasks"],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:StopTask"],
              resources: [taskArn, taskDefArn],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["dynamodb:UpdateItem"],
              resources: [props.scenariosTable.tableArn],
            }),
          ],
        }),
      },
    });
    taskCancelerRole.attachInlinePolicy(props.cloudWatchLogsPolicy);

    const taskCancelerRoleResource = taskCancelerRole.node.defaultChild as CfnResource;
    taskCancelerRoleResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "ecs:ListTasks does not support resource level permissions",
        },
      ],
    });

    this.taskCanceler = new LambdaFunction(this, "TaskCanceler", {
      description: "Stops ECS task",
      handler: "index.handler",
      role: taskCancelerRole,
      code: Code.fromBucket(props.sourceCodeBucket, `${props.sourceCodePrefix}/task-canceler.zip`),
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(300),
      environment: {
        METRIC_URL: props.metricsUrl,
        SOLUTION_ID: props.solutionId,
        VERSION: props.solutionVersion,
        SCENARIOS_TABLE: props.scenariosTable.tableName,
      },
    });
    const taskCancelerResource = this.taskCanceler.node.defaultChild as CfnResource;
    taskCancelerResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "This Lambda function does not require a VPC",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });

    this.taskCancelerInvokePolicy = new Policy(this, "TaskCancelerInvokePolicy", {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ["lambda:InvokeFunction"],
          resources: [this.taskCanceler.functionArn],
        }),
      ],
    });

    const taskStatusCheckerRole = new Role(this, "TaskStatusRole", {
      assumedBy: new ServicePrincipal("lambda.amazonaws.com"),
      inlinePolicies: {
        TaskStatusPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:ListTasks"],
              resources: ["*"],
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["ecs:DescribeTasks"],
              resources: [taskArn],
            }),
          ],
        }),
      },
    });
    taskStatusCheckerRole.attachInlinePolicy(this.taskCancelerInvokePolicy);
    taskStatusCheckerRole.attachInlinePolicy(props.cloudWatchLogsPolicy);
    taskStatusCheckerRole.attachInlinePolicy(props.scenariosDynamoDbPolicy);

    const taskStatusCheckerRoleResource = taskStatusCheckerRole.node.defaultChild as CfnResource;
    taskStatusCheckerRoleResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W11",
          reason: "ecs:ListTasks does not support resource level permissions",
        },
      ],
    });

    this.taskStatusChecker = new LambdaFunction(this, "TaskStatusChecker", {
      description: "Task status checker",
      handler: "index.handler",
      role: taskStatusCheckerRole,
      code: Code.fromBucket(props.sourceCodeBucket, `${props.sourceCodePrefix}/task-status-checker.zip`),
      runtime: Runtime.NODEJS_16_X,
      timeout: Duration.seconds(180),
      environment: {
        SCENARIOS_TABLE: props.scenariosTable.tableName,
        TASK_CANCELER_ARN: this.taskCanceler.functionArn,
        SOLUTION_ID: props.solutionId,
        VERSION: props.solutionVersion,
      },
    });
    const taskStatusCheckerResource = this.taskStatusChecker.node.defaultChild as CfnResource;
    taskStatusCheckerResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W58",
          reason: "CloudWatchLogsPolicy covers a permission to write CloudWatch logs.",
        },
        {
          id: "W89",
          reason: "This Lambda function does not require a VPC",
        },
        {
          id: "W92",
          reason: "Does not run concurrent executions",
        },
      ],
    });
  }
}
