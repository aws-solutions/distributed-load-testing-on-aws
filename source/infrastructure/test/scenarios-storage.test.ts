// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { App, DefaultStackSynthesizer, Stack } from "aws-cdk-lib";
import { ScenarioTestRunnerStorageConstruct } from "../lib/back-end/scenarios-storage";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { createTemplateWithoutS3Key } from "./snapshot_helpers";

test("DLT API Test", () => {
  const app = new App();
  const stack = new Stack(app, "DLTStack", {
    synthesizer: new DefaultStackSynthesizer({
      generateBootstrapVersionRule: false,
    }),
  });
  const testLogsBucket = new Bucket(stack, "testLogsBucket");

  const storage = new ScenarioTestRunnerStorageConstruct(stack, "TestScenarioStorage", {
    s3LogsBucket: testLogsBucket,
    webAppURL: "test.exampledomain.com",
    solutionId: "testId",
  });

  expect(createTemplateWithoutS3Key(stack)).toMatchSnapshot();
  expect(storage.scenariosBucket).toBeDefined();
  expect(storage.scenariosS3Policy).toBeDefined();
  expect(storage.scenariosTable).toBeDefined();
  expect(storage.historyTable).toBeDefined();
  expect(storage.scenarioDynamoDbPolicy).toBeDefined();
  expect(storage.historyDynamoDbPolicy).toBeDefined();
});
