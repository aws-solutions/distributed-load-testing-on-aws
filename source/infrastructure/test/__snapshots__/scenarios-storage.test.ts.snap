// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT API Test 1`] = `
{
  "Resources": {
    "TestScenarioStorageDLTHistoryTable3639137E": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "testId",
            "AttributeType": "S",
          },
          {
            "AttributeName": "testRunId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "testId",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "testRunId",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "TestScenarioStorageDLTScenariosBucket9A78F6FF": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "CorsConfiguration": {
          "CorsRules": [
            {
              "AllowedHeaders": [
                "*",
              ],
              "AllowedMethods": [
                "GET",
                "POST",
                "PUT",
              ],
              "AllowedOrigins": [
                "https://test.exampledomain.com",
              ],
              "ExposedHeaders": [
                "ETag",
              ],
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "AbortIncompleteMultipartUpload": {
                "DaysAfterInitiation": 7,
              },
              "ExpiredObjectDeleteMarker": true,
              "Status": "Enabled",
            },
          ],
        },
        "LoggingConfiguration": {
          "DestinationBucketName": {
            "Ref": "testLogsBucket85E419AD",
          },
          "LogFilePrefix": "scenarios-bucket-access/",
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestScenarioStorageDLTScenariosBucketPolicyDD7EC971": {
      "Properties": {
        "Bucket": {
          "Ref": "TestScenarioStorageDLTScenariosBucket9A78F6FF",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestScenarioStorageDLTScenariosBucket9A78F6FF",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestScenarioStorageDLTScenariosBucket9A78F6FF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": false,
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TestScenarioStorageDLTScenariosBucket9A78F6FF",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "TestScenarioStorageDLTScenariosBucket9A78F6FF",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "TestScenarioStorageDLTScenariosTable136C8D56": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "testId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "testId",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "testLogsBucket85E419AD": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
